### PR TITLE
Kmeans working example

### PR DESCRIPTION
**Pull request by Hzaatiti**
Kmeans example working on the Poincare manifold and the Hypersphere,

I make a seperate mean function conditioned by "mean method", should be embedded in the exisitng mean computation loop later on.

### CODE REVIEW

File: geomstats/backend/pytorch/__init__.py [hunk 1]
```diff
@@ -235,7 +235,8 @@ def to_ndarray(x, to_ndim, axis=0):
     x = array(x)
     if x.dim() == to_ndim - 1:
         x = torch.unsqueeze(x, dim=axis)
-    assert x.dim() >= to_ndim
+
+# assert x.dim() >= to_ndim
```
Comment (ninamiolane) [inline geomstats/backend/pytorch/__init__.py]: Remove left-over comment

File: geomstats/backend/numpy/__init__.py [hunk 2]
```diff
@@ -158,12 +159,14 @@ def to_ndarray(x, to_ndim, axis=0):
...
-    assert x.ndim >= to_ndim
+
+    if x.ndim != 0:
+        assert x.ndim >= to_ndim
     return x
 
 
-def norm(val, axis):
-    return _np.linalg.norm(val, axis=axis)
+def norm(val, axis, keepdims=False):
```
Comment (ninamiolane) [inline geomstats/backend/numpy/__init__.py]: This should be in the file `/numpy/linalg.py` and called as `gs.linalg.norm` in the code.

File: geomstats/geometry/hyperbolic.py [hunk 3]
```diff
@@ -667,15 +667,17 @@ def log(self, point, base_point):
 
             norm_add = gs.to_ndarray(gs.norm(add_base_point, -1), 2, -1)
             norm_add = gs.repeat(norm_add, base_point.shape[-1], -1)
-            norm2_base_point = gs.to_ndarray(gs.sum(base_point**2, -1), 2, -1)
+
+            norm2_base_point = gs.to_ndarray(gs.norm(base_point, -1), 2, -1)
```
Comment (ninamiolane) [inline geomstats/geometry/hyperbolic.py]: use `gs.linalg.norm`

File: geomstats/geometry/hyperbolic.py [hunk 4]
```diff
@@ -667,15 +667,17 @@ def log(self, point, base_point):
 
             norm_add = gs.to_ndarray(gs.norm(add_base_point, -1), 2, -1)
```
Comment (ninamiolane) [inline geomstats/geometry/hyperbolic.py]: use gs.linalg.norm

File: geomstats/geometry/lie_algebra.py [hunk 5]
```diff
@@ -10,7 +10,7 @@
 
 BCH_INFO = gs.asarray([
     [int(x) for x in i.strip().split()]
-    for i in open("geomstats/geometry/bch_coefficients.dat").readlines()
+    for i in open("/geomstats/geometry/bch_coefficients.dat").readlines()
```
Comment (ninamiolane) [inline geomstats/geometry/lie_algebra.py]: Remove the `/`, we need a relative path here, otherwise it is dependent on the user's file organization.

File: geomstats/geometry/riemannian_metric.py [hunk 6]
```diff
@@ -350,80 +350,113 @@ def mean(self, points,
 
         verbose: bool, optional
         """
+        if point_type != 'ball':
```
Comment (ninamiolane) [inline geomstats/geometry/riemannian_metric.py]: How does your Frechet mean algorithm differ from the one implemented here? It seems that the stopping condition and the learning rate are the main differences, is there something else?

We should not distinguish the different cases of the Frechet mean computation by deciding by a condition on point_type. Instead:
- create another private method with your frechet mean algorithm
- call this specific method in your example
or:
- address issue #364 by creating a wrapper mean that decides which frechet mean algorithm to call based on an additional parameter called `method`, for example.

File: geomstats/geometry/riemannian_metric.py [hunk 7]
```diff
@@ -350,80 +350,113 @@ def mean(self, points,
...
+                    last_iteration, variance, sq_dist))
+
+            mean = gs.to_ndarray(mean, to_ndim=2)
             return mean
 
-        sq_dists_between_iterates = []
-        iteration = 0
-        sq_dist = gs.array([[0.]])
-        variance = gs.array([[0.]])
+        if point_type == 'ball':
```
Comment (ninamiolane) [inline geomstats/geometry/riemannian_metric.py]: What follows should go into a new mean method.

File: geomstats/learning/kmeans.py [hunk 8]
```diff
@@ -73,8 +74,12 @@ def fit(self, X, max_iter=100):
             old_centroids = gs.copy(self.centroids)
             for i in range(self.n_clusters):
                 fold = gs.squeeze(X[belongs == i])
+
                 if len(fold) > 0:
-                    self.centroids[i] = self.riemannian_metric.mean(fold)
+                    self.centroids[i] = self.riemannian_metric.mean(
```
Comment (ninamiolane) [inline geomstats/learning/kmeans.py]: Here, you would call your mean method

File: geomstats/learning/kmeans.py [hunk 9]
```diff
@@ -38,6 +38,7 @@ def __init__(self, riemannian_metric, n_clusters=8, init='random',
         self.riemannian_metric = riemannian_metric
         self.tol = tol
         self.verbose = verbose
+        self.point_type = point_type
```
Comment (ninamiolane) [inline geomstats/learning/kmeans.py]: Adding point_type as an attribute may not be necessary once you have created your mean method

File: tests/test_hyperbolic.py [hunk 10]
```diff
@@ -17,7 +17,7 @@
 RTOL = 1e-6
 
 
-class TestHyperbolicMethods(geomstats.tests.TestCase):
+class TestHyperbolicSpaceMethods(geomstats.tests.TestCase):
```
Comment (ninamiolane) [inline tests/test_hyperbolic.py]: Rebase on master, as there has been a renaming across the database today ;)

File: geomstats/geometry/riemannian_metric.py (hunk 6 above)
Comment (Hzaatiti) [reply geomstats/geometry/riemannian_metric.py]: > How does your Frechet mean algorithm differ from the one implemented here? It seems that the stopping condition and the learning rate are the main differences, is there something else?

There's something with the inner product that is not working on the coordinates type that I am using that I've still didn't work out. But you are right, with some additional work I should be able to include the frechet mean I use with the existing loop (still working on it). So for the moment I went with your proposed solution:

> address issue #364 by creating a wrapper mean that decides which frechet mean algorithm to call based on an additional parameter called method, for example.

Which I am not sure I did so well as a wrapper, I just added an extra parameter for the moment.

File: geomstats/geometry/hyperbolic.py [hunk 11]
```diff
@@ -476,7 +477,7 @@ def random_uniform(self, n_samples=1, bound=1.):
 
 
 class HyperbolicMetric(RiemannianMetric):
-    """Class for the Hyperbolic metric."""
+    """Class that defines operations using a Riemannian metric."""
```
Comment (stefanheyder) [inline geomstats/geometry/hyperbolic.py]: Maybe replace Riemannian with Hyperbolic? The above comment is generic for Riemannian metrics, but I this class provides functions related to that specific metric

File: geomstats/geometry/hyperbolic.py [hunk 12]
```diff
@@ -488,7 +489,8 @@ def __init__(self, dimension, point_type='extrinsic', scale=1):
         self.scale = scale
 
     def inner_product(self, tangent_vec_a, tangent_vec_b, base_point=None):
-        """Compute the inner product of two tangent vectors at a base point.
+        """
```
Comment (stefanheyder) [inline geomstats/geometry/hyperbolic.py]: Please add a short description (see https://github.com/geomstats/geomstats/blob/master/docs/contributing.rst#guidelines-for-writing-documentation).

File: geomstats/geometry/hyperbolic.py [hunk 13]
```diff
@@ -670,28 +676,32 @@ def log(self, point, base_point):
...
             norm2_base_point = gs.repeat(norm2_base_point,
                                          base_point.shape[-1], -1)
 
-            log = (1 - norm2_base_point) * gs.arctanh(norm_add)\
+            log = (1 - norm2_base_point**2) * gs.arctanh(norm_add)\
                 * (add_base_point / norm_add)
 
             mask_0 = gs.all(gs.isclose(norm_add, 0))
             log[mask_0] = 0
+            log[gs.isnan(log)] = 0
```
Comment (stefanheyder) [inline geomstats/geometry/hyperbolic.py]: I'm not sure whether this is the way to go, do you know which cases might lead to a `nan` in the `log` array? I think we should rather throw an Exception or somehow mark the points where computation failed

File: geomstats/geometry/hyperbolic.py [hunk 14]
```diff
@@ -670,28 +676,32 @@ def log(self, point, base_point):
...
 
             return log
         else:
             raise NotImplementedError(
                     'log is only implemented for ball and extrinsic')
 
     def mobius_add(self, point_a, point_b):
-        """Compute the mobius addition of two points.
+        """Mobius addition of two points.
```
Comment (stefanheyder) [inline geomstats/geometry/hyperbolic.py]: Could you also add a short formula describing the mobius addition?

File: geomstats/geometry/hypersphere.py [hunk 15]
```diff
@@ -303,6 +303,40 @@ def random_von_mises_fisher(self, kappa=10, n_samples=1):
 
         return point
 
+    def random_von_mises_fisher2(self, kappa=10, n_samples=1):
```
Comment (stefanheyder) [inline geomstats/geometry/hypersphere.py]: Is there a difference between this method and the one just above (`random_von_mises_fisher`) ?

File: geomstats/geometry/lie_algebra.py [hunk 16]
```diff
@@ -104,7 +104,7 @@ def basis_representation(self, matrix_representation):
         matrix_representation: array-like, shape=[n_sample, n, n]
 
         Returns
-        -------
+        ------
```
Comment (stefanheyder) [inline geomstats/geometry/lie_algebra.py]: I think the official numpy docstring guide (https://numpydoc.readthedocs.io/en/latest/format.html) also has 7 dashes below the Returns (instead of the 6 you have here)

File: geomstats/geometry/riemannian_metric.py [hunk 17]
```diff
@@ -1,4 +1,6 @@
-"""Riemannian and pseudo-Riemannian metrics."""
+"""
+Riemannian and pseudo-Riemannian metrics.
```
Comment (stefanheyder) [inline geomstats/geometry/riemannian_metric.py]: Did you rebase your code recently? I think many of the docstring issues with your PR stem from the fact that we merged the docstrings PR from @cshewmake2 a few days ago ...

File: tests/test_hyperbolic.py [hunk 18]
```diff
@@ -446,35 +457,43 @@ def test_scaled_inner_product(self):
...
         result = inner_product_scaled_metric
         expected = scale ** 2 * inner_product_default_metric
         self.assertAllClose(result, expected)
 
     @geomstats.tests.np_only
     def test_scaled_squared_norm(self):
         base_point_intrinsic = gs.array([1, 1, 1])
-        base_point = self.space.intrinsic_to_extrinsic_coords(
-            base_point_intrinsic)
+        base_point = \
```
Comment (stefanheyder) [inline tests/test_hyperbolic.py]: Does this also stem from the docstring PR? I'd rather prefer the old formatting with the function call on the first whole line, arguments on the second line.

Comment (stefanheyder) [reply tests/test_hyperbolic.py]: Which is also what our contributing file says

File: geomstats/backend/numpy/__init__.py [hunk 19]
```diff
@@ -161,12 +162,10 @@ def to_ndarray(x, to_ndim, axis=0):
     x = _np.array(x)
     if x.ndim == to_ndim - 1:
         x = _np.expand_dims(x, axis=axis)
-    assert x.ndim >= to_ndim
-    return x
-
 
-def norm(val, axis):
```
[the remark this answers is not in the document]
Comment (Hzaatiti) [reply geomstats/backend/numpy/__init__.py]: Thanks for all your comments :) 
Yes I ran a search and I haven't came across gs.norm anywhere.

File: geomstats/backend/pytorch/__init__.py [hunk 20]
```diff
@@ -235,7 +235,6 @@ def to_ndarray(x, to_ndim, axis=0):
     x = array(x)
     if x.dim() == to_ndim - 1:
         x = torch.unsqueeze(x, dim=axis)
-    assert x.dim() >= to_ndim
```
[the remark this answers is not in the document]
Comment (Hzaatiti) [reply geomstats/backend/pytorch/__init__.py]: Yes I ran across some errors in some situations were x was empty if i remember right.

File: geomstats/geometry/hyperbolic.py [hunk 21]
```diff
@@ -531,7 +533,8 @@ def squared_norm(self, vector, base_point=None):
         return sq_norm
 
     def exp(self, tangent_vec, base_point):
-        """Compute Riemannian exponential of tangent vector wrt to base point.
+        """
+        Riemannian exponential of a tangent vector wrt to a base point.
```
[the remark this answers is not in the document]
Comment (Hzaatiti) [reply geomstats/geometry/hyperbolic.py]: true! i wasn't using the numpy docstring convention, fixed

File: geomstats/geometry/hypersphere.py (hunk 15 above)
Comment (Hzaatiti) [reply geomstats/geometry/hypersphere.py]: oh sorry was doing some tests and forgot to remove it. thanks

File: geomstats/geometry/hyperbolic.py (hunk 14 above)
Comment (Hzaatiti) [reply geomstats/geometry/hyperbolic.py]: Added formula using .. math:: operator. Not sure how to test if that would work well.

File: geomstats/geometry/hyperbolic.py (hunk 13 above)
Comment (Hzaatiti) [reply geomstats/geometry/hyperbolic.py]: True, I'll take a closer look.

File: geomstats/geometry/riemannian_metric.py [hunk 22]
```diff
@@ -295,39 +253,35 @@ def point_on_geodesic(t):
         return point_on_geodesic
 
     def squared_dist(self, point_a, point_b):
-        """Compute squared geodesic distance between two points.
+        """
```
Comment (ninamiolane) [inline geomstats/geometry/riemannian_metric.py]: Docstring formatting does not seem right: did you rebase on master?

File: geomstats/geometry/riemannian_metric.py [hunk 23]
```diff
@@ -338,16 +292,14 @@ def variance(self,
                  weights=None,
                  base_point=None,
                  point_type='vector'):
-        """Compute variance of (weighted) points wrt a base point.
+        """
```
Comment (ninamiolane) [inline geomstats/geometry/riemannian_metric.py]: same here

File: geomstats/geometry/riemannian_metric.py [hunk 24]
```diff
@@ -550,21 +544,10 @@ def adaptive_gradientdescent_mean(self, points,
         return gs.to_ndarray(current_mean, to_ndim=2)
 
     def tangent_pca(self, points, base_point=None, point_type='vector'):
-        """Perform tPCA of points on the tangent space at a base point.
+        """Tangent PCA of points on the tangent space wrt a base point.
```
Comment (ninamiolane) [inline geomstats/geometry/riemannian_metric.py]: Docstring problem

File: geomstats/geometry/hyperbolic.py (hunk 13 above)
Comment (ninamiolane) [reply geomstats/geometry/hyperbolic.py]: Any update on this?

**Comment by ninamiolane** [comment]
Also for next time, if you want to speed up the process: smaller PR take less time to be merged ;)
